File: internal/config/config_test.go
package config

import "testing"

// testdata/nix-rendered-example.json is not hand-written: it's the actual
// config.json modules/core.nix + modules/netbird-provider.nix rendered for
// this repo's own README quickstart example, captured via `nix eval` against
// a minimal evalModules harness during development (see the deviations and
// clarifications documented in docs/providers.md — this fixture is what
// pins them down as actually-agreeing, not just independently plausible).
// If the Nix schema and this Go schema ever drift apart, this test is what
// catches it.
func TestLoad_NixRenderedExample(t *testing.T) {
	cfg, err := Load("testdata/nix-rendered-example.json")
	if err != nil {
		t.Fatalf("Load: %v", err)
	}

	peer, ok := cfg.Peers["host-b"]
	if !ok {
		t.Fatalf("expected peer %q in rendered config", "host-b")
	}
	if len(peer.Hostnames) != 1 || peer.Hostnames[0] != "host-b" {
		t.Errorf("peer host-b hostnames = %v, want [host-b]", peer.Hostnames)
	}
	if len(peer.Transports) != 2 {
		t.Fatalf("peer host-b: got %d transports, want 2 (one hand-written, one netbird-provider-contributed)", len(peer.Transports))
	}

	var sawNetbird, sawTCP bool
	for _, tr := range peer.Transports {
		switch tr.ProviderID {
		case "netbird":
			sawNetbird = true
			if tr.Probe.Method != "exec" {
				t.Errorf("netbird transport: probe.method = %q, want exec", tr.Probe.Method)
			}
			// probe.exec must be a full command line whose argv[0] is an
			// absolute Nix store path (see the "Deviation: probe.exec is
			// a command line" note in docs/providers.md); the actual
			// tokenizing/exec'ing of it is internal/probe's concern and
			// is covered by internal/probe's own tests.
			if tr.Probe.Exec == "" || tr.Probe.Exec[0] != '/' {
				t.Errorf("netbird transport: probe.exec = %q, want a command line starting with an absolute path", tr.Probe.Exec)
			}
		case "":
			sawTCP = true
			if tr.Probe.Method != "tcp" {
				t.Errorf("hand-written transport: probe.method = %q, want tcp", tr.Probe.Method)
			}
			if tr.EffectiveTarget() != "192.0.2.20" {
				t.Errorf("hand-written transport: EffectiveTarget() = %q, want 192.0.2.20", tr.EffectiveTarget())
			}
		}
	}
	if !sawNetbird {
		t.Errorf("never saw the netbird-provider-contributed transport")
	}
	if !sawTCP {
		t.Errorf("never saw the hand-written tcp transport")
	}

	uplink, ok := cfg.Uplinks["internet"]
	if !ok {
		t.Fatalf("expected uplink %q in rendered config", "internet")
	}
	if len(uplink.Transports) != 2 {
		t.Fatalf("uplink internet: got %d transports, want 2", len(uplink.Transports))
	}
	for _, tr := range uplink.Transports {
		if tr.Interface == "" {
			t.Errorf("uplink transport missing interface: %+v", tr)
		}
	}
	if !uplink.Publish.RouteMetric {
		t.Errorf("uplink internet: publish.routeMetric = false, want true")
	}
	if uplink.Publish.MetricBase != 100 || uplink.Publish.MetricStep != 50 {
		t.Errorf("uplink internet: metricBase/metricStep = %d/%d, want 100/50", uplink.Publish.MetricBase, uplink.Publish.MetricStep)
	}

	if cfg.NeedsNetAdmin() != true {
		t.Errorf("NeedsNetAdmin() = false, want true (routeMetric is enabled)")
	}
	if cfg.NeedsNetRaw() != true {
		t.Errorf("NeedsNetRaw() = false, want true (wireless0 uses icmp+bindToInterface)")
	}
}

File: tests/config_test.rs
//! testdata/nix-rendered-example.json is not hand-written: it's the
//! actual config.json `modules/core.nix` + `modules/netbird-provider.nix`
//! rendered for this repo's own README quickstart example, captured via
//! `nix eval` against a minimal evalModules harness during development
//! (see the deviations and clarifications documented in
//! docs/providers.md -- this fixture is what pins them down as
//! actually-agreeing, not just independently plausible). If the Nix
//! schema and this Rust schema ever drift apart, this test is what
//! catches it.

use std::path::Path;

use nixnet::config;

#[test]
fn load_nix_rendered_example() {
    let cfg = config::load(Path::new("tests/testdata/nix-rendered-example.json")).expect("load");

    let peer = cfg
        .peers
        .get("host-b")
        .expect("expected peer \"host-b\" in rendered config");
    assert_eq!(peer.hostnames, vec!["host-b".to_string()]);
    assert_eq!(
        peer.transports.len(),
        2,
        "peer host-b: want 2 transports (one hand-written, one netbird-provider-contributed)"
    );

    let mut saw_netbird = false;
    let mut saw_tcp = false;
    for tr in &peer.transports {
        match tr.provider_id.as_str() {
            "netbird" => {
                saw_netbird = true;
                assert_eq!(tr.probe.method, "exec");
                // probe.exec must be a full command line whose argv[0] is
                // an absolute Nix store path (see the "Deviation:
                // probe.exec is a command line" note in
                // docs/providers.md); the actual tokenizing/exec'ing of
                // it is crate::probe's own concern, covered by its own
                // tests.
                assert!(
                    !tr.probe.exec.is_empty() && tr.probe.exec.starts_with('/'),
                    "netbird transport: probe.exec = {:?}, want a command line starting with an absolute path",
                    tr.probe.exec
                );
            }
            "" => {
                saw_tcp = true;
                assert_eq!(tr.probe.method, "tcp");
                assert_eq!(tr.effective_target(), "192.0.2.20");
            }
            other => panic!("unexpected providerId {:?}", other),
        }
    }
    assert!(
        saw_netbird,
        "never saw the netbird-provider-contributed transport"
    );
    assert!(saw_tcp, "never saw the hand-written tcp transport");

    let uplink = cfg
        .uplinks
        .get("internet")
        .expect("expected uplink \"internet\" in rendered config");
    assert_eq!(
        uplink.transports.len(),
        2,
        "uplink internet: want 2 transports"
    );
    for tr in &uplink.transports {
        assert!(
            !tr.interface.is_empty(),
            "uplink transport missing interface: {:?}",
            tr
        );
    }
    assert!(
        uplink.publish.route_metric,
        "uplink internet: publish.routeMetric = false, want true"
    );
    assert_eq!(uplink.publish.metric_base, 100);
    assert_eq!(uplink.publish.metric_step, 50);

    assert!(
        cfg.needs_net_admin(),
        "needs_net_admin() = false, want true (routeMetric is enabled)"
    );
    assert!(
        cfg.needs_net_raw(),
        "needs_net_raw() = false, want true (wireless0 uses icmp+bindToInterface)"
    );

    // STALE-2: the bound the README quickstart declares has to arrive
    // intact. A bound that renders but does not reach the daemon is
    // indistinguishable from no bound at all -- which is the defect.
    assert_eq!(
        peer.last_known_good.max_age_sec,
        Some(900),
        "the rendered lastKnownGood.maxAgeSec did not survive the wire"
    );
}

/// The other shape the same field takes: Nix renders an UNSET bound as an
/// explicit JSON `null`, which the daemon must read as "unbounded" rather
/// than as a parse error. This is what every peer that has not opted in
/// emits, so getting it wrong would fail closed on the majority case.
#[test]
fn an_unset_last_known_good_bound_reads_as_unbounded() {
    let dir = tempfile::tempdir().unwrap();
    let path = dir.path().join("config.json");
    std::fs::write(
        &path,
        r#"{
          "peers": {
            "host-b": {
              "hostnames": ["host-b"],
              "lastKnownGood": { "maxAgeSec": null },
              "transports": [ { "address": "192.0.2.20", "priority": 10 } ]
            }
          }
        }"#,
    )
    .unwrap();

    let cfg = config::load(&path).expect("an explicit null must load, not error");
    assert_eq!(cfg.peers["host-b"].last_known_good.max_age_sec, None);
}

/// A zero or negative bound is not "expire immediately" -- `onAllDown =
/// "unpublish"` is. Nix's `ints.positive` rejects it at eval time; a
/// hand-written config.json (a supported adoption path) reaches the daemon
/// unchecked, so the daemon checks it too rather than silently reading a
/// typo as a policy change.
#[test]
fn a_non_positive_last_known_good_bound_is_rejected() {
    let dir = tempfile::tempdir().unwrap();
    let path = dir.path().join("config.json");
    std::fs::write(
        &path,
        r#"{
          "peers": {
            "host-b": {
              "hostnames": ["host-b"],
              "lastKnownGood": { "maxAgeSec": 0 },
              "transports": [ { "address": "192.0.2.20", "priority": 10 } ]
            }
          }
        }"#,
    )
    .unwrap();

    let err = config::load(&path).expect_err("maxAgeSec = 0 must not load");
    let msg = err.to_string();
    assert!(
        msg.contains("maxAgeSec"),
        "error does not name the offending option: {msg}"
    );
}
